File: android/app/src/main/java/com/nextpage/data/remote/sync/ConflictResolver.kt
package com.nextpage.data.remote.sync

interface ConflictResolver<T> {
    fun resolve(local: T?, remote: T): T
}

interface VersionedSyncRecord {
    val recordId: String
    val updatedAtEpochMillis: Long
    val deletedAtEpochMillis: Long?
}

/**
 * LWW resolver unified for all 5 tables:
 * reading_progress, highlights, bookmarks, reading_sessions, user_books.
 * Rule: remote.updatedAt > local → remote; < → local; == → recordId
 * lexicographic (remote > local → remote else local). Tombstone:
 * deletedAt comparison as above — any deleted beats non-deleted, both
 * deleted → later deletedAt wins, tie → local.
 */
class LastWriteWinsConflictResolver<T : VersionedSyncRecord> : ConflictResolver<T> {
    override fun resolve(local: T?, remote: T): T {
        val localRecord = local ?: return remote

        if (localRecord.deletedAtEpochMillis != null || remote.deletedAtEpochMillis != null) {
            return resolveTombstone(localRecord, remote)
        }

        return chooseLatest(localRecord, remote)
    }

    private fun resolveTombstone(local: T, remote: T): T {
        val localDeletedAt = local.deletedAtEpochMillis
        val remoteDeletedAt = remote.deletedAtEpochMillis

        if (localDeletedAt != null && remoteDeletedAt == null) return local
        if (localDeletedAt == null && remoteDeletedAt != null) return remote
        if (localDeletedAt != null && remoteDeletedAt != null) {
            return if (localDeletedAt >= remoteDeletedAt) local else remote
        }
        return chooseLatest(local, remote)
    }

    private fun chooseLatest(local: T, remote: T): T {
        if (remote.updatedAtEpochMillis > local.updatedAtEpochMillis) return remote
        if (remote.updatedAtEpochMillis < local.updatedAtEpochMillis) return local
        return if (remote.recordId > local.recordId) remote else local
    }
}

File: desktop/src/lib/shared/sync/ConflictResolver.ts
import type { VersionedSyncRecord } from '../services/storage/StorageProvider';

export interface ConflictResolver<T> {
  resolve(local: T | null | undefined, remote: T): T;
}

export class LastWriteWinsConflictResolver<T extends VersionedSyncRecord> implements ConflictResolver<T> {
  resolve(local: T | null | undefined, remote: T): T {
    if (!local) return remote;

    if (local.deletedAtEpochMillis !== null || remote.deletedAtEpochMillis !== null) {
      return this.resolveTombstone(local, remote);
    }

    return this.chooseLatest(local, remote);
  }

  private resolveTombstone(local: T, remote: T): T {
    const localDeletedAt = local.deletedAtEpochMillis;
    const remoteDeletedAt = remote.deletedAtEpochMillis;

    if (localDeletedAt !== null && remoteDeletedAt === null) return local;
    if (localDeletedAt === null && remoteDeletedAt !== null) return remote;
    if (localDeletedAt !== null && remoteDeletedAt !== null) {
      return localDeletedAt >= remoteDeletedAt ? local : remote;
    }
    return this.chooseLatest(local, remote);
  }

  private chooseLatest(local: T, remote: T): T {
    if (remote.updatedAtEpochMillis > local.updatedAtEpochMillis) return remote;
    if (remote.updatedAtEpochMillis < local.updatedAtEpochMillis) return local;
    return remote.recordId > local.recordId ? remote : local;
  }
}
